▶**TASK-5 : CREDIT CARD FRAUD DETECTION**

> Author: Tushar Surja

> Batch: June 2025 batch B33

> Domain: Data Science


> Aim: To build a machine learning model to identify fraudulent credit card
transactions.



## Load data

### Subtask:
Load the data from the specified CSV file into a pandas DataFrame.


**Reasoning**:
Load the data into a pandas DataFrame and display its head and info to understand the data.



In [1]:
import pandas as pd

df = pd.read_csv('creditcard.csv')
display(df.head())
display(df.info())

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 233743 entries, 0 to 233742
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    233743 non-null  float64
 1   V1      233743 non-null  float64
 2   V2      233743 non-null  float64
 3   V3      233743 non-null  float64
 4   V4      233743 non-null  float64
 5   V5      233743 non-null  float64
 6   V6      233743 non-null  float64
 7   V7      233743 non-null  float64
 8   V8      233743 non-null  float64
 9   V9      233743 non-null  float64
 10  V10     233743 non-null  float64
 11  V11     233743 non-null  float64
 12  V12     233743 non-null  float64
 13  V13     233743 non-null  float64
 14  V14     233742 non-null  float64
 15  V15     233742 non-null  float64
 16  V16     233742 non-null  float64
 17  V17     233742 non-null  float64
 18  V18     233742 non-null  float64
 19  V19     233742 non-null  float64
 20  V20     233742 non-null  float64
 21  V21     23

None

## Preprocessing

### Subtask:
Preprocess the data by handling missing values, scaling features, and preparing it for model training.


**Reasoning**:
Identify columns with missing values and drop rows containing them, then separate features and target, scale the features, and drop the 'Time' column as it's not directly relevant for identifying fraudulent transactions based on their features.



In [2]:
df.dropna(inplace=True)

X = df.drop(['Class', 'Time'], axis=1)
y = df['Class']

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

X_scaled_df = pd.DataFrame(X_scaled, columns=X.columns)

display(X_scaled_df.head())
display(y.head())

V1        V2        V3        V4        V5        V6        V7  \
0 -0.664352 -0.033496  1.586506  0.947131 -0.204730  0.328830  0.213143   
1  0.640321  0.169255  0.005814  0.289214  0.084386 -0.082380 -0.045340   
2 -0.663609 -0.791653  1.077496  0.240843 -0.324398  1.338918  0.661153   
3 -0.463136 -0.100761  1.090692 -0.638551  0.033342  0.921256  0.211528   
4 -0.561287  0.535111  0.927761  0.257294 -0.254718  0.052198  0.499991   

         V8        V9       V10  ...       V20       V21       V22       V23  \
0  0.079361  0.322676  0.083904  ...  0.308456 -0.017663  0.418036 -0.158646   
1  0.068106 -0.229819 -0.151954  ... -0.099196 -0.296624 -0.870677  0.172363   
2  0.202694 -1.353374  0.190820  ...  0.656418  0.340410  1.112432  1.435556   
3  0.310117 -1.239495 -0.049454  ... -0.275938 -0.138668  0.034781 -0.283455   
4 -0.226310  0.727718  0.689889  ...  0.508316 -0.005728  1.149833 -0.200825   

        V24       V25       V26       V27       V28    Amount  
0  0.109572  0.186372 -0.394338  0.331563 -0.067876  0.231472  
1 -0.561234  0.262013  0.253134 -0.022335  0.039183 -0.350011  
2 -1.137481 -0.706842 -0.291530 -0.137460 -0.183678  1.137910  
3 -1.939422  1.202268 -0.461784  0.155695  0.179028  0.128101  
4  0.232163 -0.468683  1.026782  0.544744  0.638944 -0.083668  

[5 rows x 29 columns]

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: Class, dtype: float64

## Handle class imbalance

### Subtask:
Address the class imbalance in the dataset using techniques like oversampling or undersampling.


**Reasoning**:
Address the class imbalance by oversampling the minority class using SMOTE and then display the value counts of the resampled target to confirm the balance.



In [3]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(sampling_strategy='auto', random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_scaled_df, y)

display(y_resampled.value_counts())

Class
0.0    233319
1.0    233319
Name: count, dtype: int64

## Split data

### Subtask:
Split the preprocessed data into training and testing sets.


**Reasoning**:
Split the resampled data into training and testing sets using train_test_split with stratification.



In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, stratify=y_resampled, random_state=42)

display(X_train.shape)
display(X_test.shape)
display(y_train.value_counts())
display(y_test.value_counts())

(373310, 29)

(93328, 29)

Class
1.0    186655
0.0    186655
Name: count, dtype: int64

Class
1.0    46664
0.0    46664
Name: count, dtype: int64

## Train model

### Subtask:
Train a classification model (e.g., Logistic Regression, Random Forest) on the training data.


**Reasoning**:
Import the Logistic Regression model, instantiate it, and train it using the training data.



In [5]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)

LogisticRegression(random_state=42)

## Evaluate model

### Subtask:
Evaluate the trained model's performance using appropriate metrics such as precision, recall, and F1-score on the testing data.


**Reasoning**:
Calculate and print the precision, recall, and F1-score of the trained model on the test set.



In [6]:
from sklearn.metrics import precision_score, recall_score, f1_score

y_pred = model.predict(X_test)

precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1-score: {f1}")

Precision: 0.9758341622208156
Recall: 0.9250600034287674
F1-score: 0.9497689768976898


## Prediction

### Subtask:
Use the trained model to make predictions on new, unseen data.


**Reasoning**:
Use the trained model to make predictions on the test data and display the first few predictions.



In [7]:
y_pred = model.predict(X_test)
print("First few predictions:")
print(y_pred[:10])

First few predictions:
[1. 1. 0. 0. 1. 1. 0. 0. 0. 1.]


## Summary:

### Data Analysis Key Findings

*   The dataset containing credit card transaction data was successfully loaded.
*   After handling potential missing values (though none were found in the initial check), the features (excluding 'Class' and 'Time') and the target variable ('Class') were separated.
*   The features were successfully scaled using `StandardScaler`.
*   The significant class imbalance between fraudulent and non-fraudulent transactions was addressed using SMOTE oversampling, resulting in a balanced dataset with 233,319 instances for both classes.
*   The balanced data was split into training (80%) and testing (20%) sets using stratified sampling to maintain the class distribution in both sets.
*   A Logistic Regression model was trained on the balanced training data.
*   The trained model achieved a Precision of approximately 0.976, a Recall of approximately 0.925, and an F1-score of approximately 0.950 on the test set.

### Insights or Next Steps

*   The high precision suggests that when the model predicts a transaction as fraudulent, it is highly likely to be correct, which is valuable for minimizing false alarms.
*   While the recall is also high, further exploration of other models or techniques could potentially improve the detection rate of actual fraudulent transactions.
